### PR TITLE
Introduce static analysis and clean up code to pass

### PR DESCRIPTION
**Pull request by bmeike**
Major refactor to get code up to basic Java standards.

Added static analysis tools checkstyle, findbugs, lint, and pmd.

Fixed code to pass all tools

There is some evidence that I've introduced a bug here: 
```
hbase.lite.tes: JNI ERROR (app bug): attempt to use stale Global 0x25e6 (should be 0x25ea)
hbase.lite.tes: java_vm_ext.cc:542] JNI DETECTED ERROR IN APPLICATION: use of deleted global reference 0x25e6
hbase.lite.tes: java_vm_ext.cc:542]     from void com.couchbase.lite.internal.core.C4Replicator.free(
                                             long, java.lang.Object, java.lang.Object)
```
It occurs infrequently, and I am having trouble reproducing it.  Since this is such a huge commit I want to get it in and fix this and any other regressions, later.

### CODE REVIEW

File: shared/src/main/java/com/couchbase/lite/SelectResult.java [hunk 1]
```diff
@@ -132,27 +116,43 @@ private SelectResult(Expression expression) {
...
-        return new SelectResult.From(expr);
+        return new SelectResult.From(PropertyExpression.allFrom(null));
+    }
+
+    //---------------------------------------------
+    // API - public static methods
+    //---------------------------------------------
+    //---------------------------------------------
+    // member variables
+    //---------------------------------------------
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/SelectResult.java]: These section comments are not valid anymore.

File: shared/src/main/java/com/couchbase/lite/internal/core/C4BlobKey.java [hunk 2]
```diff
@@ -26,14 +26,32 @@
...
  */
 public class C4BlobKey {
+    /**
+     * Decode a string of the form "shar1-"+base64 into a raw key
+     */
+    static native long fromString(String str) throws LiteCoreException;
+
     //-------------------------------------------------------------------------
-    // Member Variables
+    // Constructor
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/core/C4BlobKey.java]: This comment is not valid.

File: shared/src/main/java/com/couchbase/lite/internal/core/C4BlobKey.java [hunk 3]
```diff
@@ -26,14 +26,32 @@
...
+    static native String toString(long blobKey);
+
+    /**
+     * Release C4BlobKey
+     */
+    static native void free(long blobKey);
 
     //-------------------------------------------------------------------------
-    // Constructor
+    // public methods
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/core/C4BlobKey.java]: This comment is not valid.

File: shared/src/main/java/com/couchbase/lite/internal/core/C4BlobKey.java [hunk 4]
```diff
@@ -59,47 +80,25 @@ public String toString() {
         return toString(handle);
     }
 
+    //-------------------------------------------------------------------------
+    // native methods
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/core/C4BlobKey.java]: This comment is not valid

File: shared/src/main/java/com/couchbase/lite/internal/core/C4BlobReadStream.java [hunk 5]
```diff
@@ -25,24 +25,33 @@
...
 public class C4BlobReadStream {
     //-------------------------------------------------------------------------
-    // Member Variables
+    // native methods
     //-------------------------------------------------------------------------
-    private long handle = 0L; // hold pointer to C4BlobReadStream
+    static native byte[] read(long readStream, long maxBytesToRead) throws LiteCoreException;
 
     //-------------------------------------------------------------------------
     // Constructor
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/core/C4BlobReadStream.java]: This comment is not valid.

File: shared/src/main/java/com/couchbase/lite/internal/core/C4BlobReadStream.java [hunk 6]
```diff
@@ -25,24 +25,33 @@
...
 
-    C4BlobReadStream(long handle) {
-        if (handle == 0)
-            throw new IllegalArgumentException("handle is 0");
-        this.handle = handle;
-    }
+    static native long getLength(long readStream) throws LiteCoreException;
 
     //-------------------------------------------------------------------------
     // public methods
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/core/C4BlobReadStream.java]: This comment is not valid.

File: shared/src/main/java/com/couchbase/lite/internal/core/C4BlobStore.java [hunk 7]
```diff
@@ -59,11 +39,53 @@
...
+    //-------------------------------------------------------------------------
+    static native long getBlobStore(long db) throws LiteCoreException;
+    //-------------------------------------------------------------------------
+    // Constructor
+    //-------------------------------------------------------------------------
+
+    static native long openStore(String dirPath, long flags) throws LiteCoreException;
+
+    //-------------------------------------------------------------------------
+    // public methods
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/core/C4BlobStore.java]: This comment is not valid.

File: shared/src/main/java/com/couchbase/lite/internal/fleece/FLValue.java [hunk 8]
```diff
@@ -17,122 +17,30 @@
...
-import static com.couchbase.lite.internal.fleece.FLConstants.FLValueType.kFLDict;
-import static com.couchbase.lite.internal.fleece.FLConstants.FLValueType.kFLNull;
-import static com.couchbase.lite.internal.fleece.FLConstants.FLValueType.kFLNumber;
-import static com.couchbase.lite.internal.fleece.FLConstants.FLValueType.kFLString;
+import com.couchbase.lite.LiteCoreException;
+
 
 public class FLValue {
     //-------------------------------------------------------------------------
     // private variables
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/internal/fleece/FLValue.java]: This comment is not valid anymore.

**Comment by bmeike** [comment]
Merging.  Will fix the broken comments in a subsequent commit, today.

**Comment by pasin** [comment]
Beside the broken comment, there is one breaking API (Database.LOG). Please also fix it as well.

File: android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java [hunk 9]
```diff
@@ -127,6 +130,30 @@ public ReplicatorConfiguration(@NonNull Database database, @NonNull Endpoint tar
     // Setters
     //---------------------------------------------
```
[the remark this answers is not in the document]
Comment (bmeike) [reply android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java]: I tried to get rid of a lot of those crazy comments.  Many of them were pretty misplaced.  I missed some.  *Additional* fixes, beyond the ones already introduced, will be part of a future commit.

[the remark this answers is not in the document]
Comment (bmeike) [reply android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java]: It looks to me as if that comment is still relevant!  The section of code below it is all setters...

**Pull request merged**